pydantic_extra_types/types/country.py

```diff
@@ -0,0 +1,1185 @@
+    'Moldova (the Republic of)': 'MD',
+    'Monaco': 'MC',
+    'Mongolia': 'MN',
+    'Montenegro': 'ME',
+    'Montserrat': 'MS',
+    'Morocco': 'MA',
+    'Mozambique': 'MZ',
+    'Myanmar': 'MM',
+    'Namibia': 'NA',
+    'Nauru': 'NR',
+    'Nepal': 'NP',
+    'Netherlands': 'NL',
+    'New Caledonia': 'NC',
+    'New Zealand': 'NZ',
+    'Nicaragua': 'NI',
+    'Niger': 'NE',
+    'Nigeria': 'NG',
+    'Niue': 'NU',
+    'Norfolk Island': 'NF',
+    'Northern Mariana Islands': 'MP',
+    'Norway': 'NO',
+    'Oman': 'OM',
+    'Pakistan': 'PK',
+    'Palau': 'PW',
+    'Panama': 'PA',
+    'Papua New Guinea': 'PG',
+    'Paraguay': 'PY',
+    'Peru': 'PE',
+    'Philippines': 'PH',
+    'Pitcairn': 'PN',
+    'Poland': 'PL',
+    'Portugal': 'PT',
+    'Puerto Rico': 'PR',
+    'Qatar': 'QA',
+    'Réunion': 'RE',
+    'Romania': 'RO',
+    'Russian Federation': 'RU',
+    'Rwanda': 'RW',
+    'Saint Barthélemy': 'BL',
+    'Saint Helena Ascension Island Tristan da Cunha': 'SH',
+    'Saint Kitts and Nevis': 'KN',
+    'Saint Lucia': 'LC',
+    'Saint Martin (French part)': 'MF',
+    'Saint Pierre and Miquelon': 'PM',
+    'Saint Vincent and the Grenadines': 'VC',
+    'Samoa': 'WS',
+    'San Marino': 'SM',
+    'Sao Tome and Principe': 'ST',
+    'Saudi Arabia': 'SA',
+    'Senegal': 'SN',
+    'Serbia': 'RS',
+    'Seychelles': 'SC',
+    'Sierra Leone': 'SL',
+    'Singapore': 'SG',
+    'Sint Maarten (Dutch part)': 'SX',
+    'Slovakia': 'SK',
+    'Slovenia': 'SI',
+    'Solomon Islands': 'SB',
+    'Somalia': 'SO',
+    'South Africa': 'ZA',
+    'South Georgia and the South Sandwich Islands': 'GS',
+    'South Sudan': 'SS',
+    'Spain': 'ES',
+    'Sri Lanka': 'LK',
+    'Sudan': 'SD',
+    'Suriname': 'SR',
+    'Svalbard Jan Mayen': 'SJ',
+    'Sweden': 'SE',
+    'Switzerland': 'CH',
+    'Syrian Arab Republic': 'SY',
+    'Taiwan (Province of China)': 'TW',
+    'Tajikistan': 'TJ',
+    'Tanzania, the United Republic of': 'TZ',
+    'Thailand': 'TH',
+    'Timor-Leste': 'TL',
+    'Togo': 'TG',
+    'Tokelau': 'TK',
+    'Tonga': 'TO',
+    'Trinidad and Tobago': 'TT',
+    'Tunisia': 'TN',
+    'Türkiye [ab]': 'TR',
+    'Turkmenistan': 'TM',
+    'Turks and Caicos Islands': 'TC',
+    'Tuvalu': 'TV',
+    'Uganda': 'UG',
+    'Ukraine': 'UA',
+    'United Arab Emirates': 'AE',
+    'United Kingdom of Great Britain and Northern Ireland': 'GB',
+    'United States Minor Outlying Islands': 'UM',
+    'United States of America': 'US',
+    'Uruguay': 'UY',
+    'Uzbekistan': 'UZ',
+    'Vanuatu': 'VU',
+    'Venezuela (Bolivarian Republic of)': 'VE',
+    'Viet Nam': 'VN',
+    'Virgin Islands (British)': 'VG',
+    'Virgin Islands (U.S.)': 'VI',
+    'Wallis and Futuna': 'WF',
+    'Western Sahara': 'EH',
+    'Yemen': 'YE',
+    'Zambia': 'ZM',
+    'Zimbabwe': 'ZW',
+}
+
+
+class CodeType(int, Enum):
+    alpha2 = 1
+    alpha3 = 2
+    numeric_code = 3
+    country_name = 4
+
+
+class Country(_repr.Representation):
```

Comment:
I'm not quite sure what you're doing here, but I think we need to do it differently :smile:.

I would suggest something like

```py
from pydantic_core import PydanticCustomError, core_schema


class CountryAlpha2(str):
    @classmethod
    def _validate(cls, value: str, **kwargs) -> bool:
        if value not in _index_by_alpha2():
            raise PydanticCustomError('country_alpha2', 'Invalid country alpha2 code')
        return value

    @classmethod
    def __get_pydantic_core_schema__(cls, **_kwargs: Any) -> core_schema.FunctionSchema:
        return core_schema.function_after_schema(
            core_schema.str_schema(to_upper=True),
            cls._validate,
            serialization=core_schema.to_string_ser_schema()
        )

    @property
    def short_name(self) -> str:
        return _index_by_alpha2()[self].short_name

    ...
```

Then you'll want to finish that, and implement `CountryAlpha3`, `CountryNumericCode`, and `CountryShortName` and `CountryOfficialName`.

I would suggest `CountryShortName` and `CountryOfficialName` use a case insensitive lookup in lax mode.

Reply:
Understood. Also, would it be wise to add a general class that accepts one of the codes(alpha2, alpha3, short, official, numeric), do a lookup, finds the matching country, and then represent it?

Reply:
I doubt it, I can't imagine a scenario where you want someone to lookup a country by any method, also people can use a union for that.

Reply:
Thanks for your CR. Would make the requested changes in the following days. Sorry for the trouble :)

Reply:
No problem at all, thanks for your help.

Sorry to request so many changes.